File: apps/ios/MobileAgent/Agent/KimiClient.swift
import Foundation

struct ModelEndpoint: Equatable {
    static let defaultBaseURL = "https://api.moonshot.cn/v1"
    static let defaultModel = "kimi-k3"
    static let `default` = try! ModelEndpoint(baseURL: defaultBaseURL, model: defaultModel)

    let baseURL: URL
    let model: String
    let host: String
    let sendsReasoningEffort: Bool

    init(baseURL value: String, model valueModel: String) throws {
        let normalizedBase = value.trimmingCharacters(in: .whitespacesAndNewlines)
            .replacingOccurrences(of: #"/+$"#, with: "", options: .regularExpression)
        let normalizedModel = valueModel.trimmingCharacters(in: .whitespacesAndNewlines)
        guard let components = URLComponents(string: normalizedBase),
              components.scheme?.lowercased() == "https",
              let host = components.host?.lowercased(), !host.isEmpty,
              components.user == nil, components.password == nil,
              components.query == nil, components.fragment == nil,
              let url = components.url else {
            throw AgentRuntimeError.operationFailed("MODEL_BASE_URL_INVALID: Base URL 必须是无凭据、query 和 fragment 的 HTTPS URL")
        }
        guard !normalizedModel.isEmpty, normalizedModel.count <= 160 else {
            throw AgentRuntimeError.operationFailed("MODEL_ID_INVALID")
        }
        baseURL = url
        model = normalizedModel
        self.host = host
        sendsReasoningEffort = host == "api.moonshot.cn"
    }
}

struct ModelEndpointStore {
    private let defaults = UserDefaults.standard

    func load() -> ModelEndpoint {
        (try? ModelEndpoint(
            baseURL: defaults.string(forKey: "model-base-url") ?? ModelEndpoint.defaultBaseURL,
            model: defaults.string(forKey: "model-id") ?? ModelEndpoint.defaultModel
        )) ?? .default
    }

    @discardableResult
    func save(baseURL: String, model: String) throws -> ModelEndpoint {
        let endpoint = try ModelEndpoint(baseURL: baseURL, model: model)
        defaults.set(endpoint.baseURL.absoluteString, forKey: "model-base-url")
        defaults.set(endpoint.model, forKey: "model-id")
        return endpoint
    }
}

@MainActor
final class OpenAICompatibleClient {
    private let apiKey: String
    let endpoint: ModelEndpoint

    init(
        apiKey: String,
        endpoint: ModelEndpoint = .default
    ) {
        self.apiKey = apiKey
        self.endpoint = endpoint
    }

    func complete(messages: [[String: Any]], tools: [[String: Any]]) async throws -> [String: Any] {
        var request = URLRequest(url: endpoint.baseURL.appending(path: "chat/completions"))
        request.httpMethod = "POST"
        request.timeoutInterval = 180
        request.setValue("Bearer \(apiKey)", forHTTPHeaderField: "Authorization")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        var body: [String: Any] = [
            "model": endpoint.model,
            "messages": messages,
            "tools": tools,
            "tool_choice": "auto",
        ]
        if endpoint.sendsReasoningEffort { body["reasoning_effort"] = "low" }
        let taskID = benchmarkTaskID(messages)
        var format: [String: Any]?
        if let taskID {
            let successfulCalls = successfulToolCalls(messages)
            let readyForFinal = taskID == "C1"
                ? hasC1CompletionEvidence(successfulCalls)
                : !successfulCalls.isEmpty
            format = readyForFinal ? responseFormat(taskID: taskID) : nil
            if let format { body["response_format"] = format }
        }
        NSLog(
            "MobileAgentPi model request endpointHost=%@ model=%@ benchmarkTask=%@ responseFormat=%@",
            endpoint.host,
            endpoint.model,
            taskID ?? "none",
            format?["type"] as? String ?? "text"
        )
        request.httpBody = try JSONSerialization.data(withJSONObject: body)

        let (data, response) = try await URLSession.shared.data(for: request)
        guard let http = response as? HTTPURLResponse else {
            throw AgentRuntimeError.operationFailed("MODEL_INVALID_RESPONSE")
        }
        let json = try JSONSerialization.jsonObject(with: data) as? [String: Any] ?? [:]
        guard (200...299).contains(http.statusCode) else {
            let error = json["error"] as? [String: Any]
            let message = error?["message"] as? String ?? "HTTP \(http.statusCode)"
            throw AgentRuntimeError.operationFailed("MODEL_API_ERROR: \(message)")
        }
        return json
    }

    private func benchmarkTaskID(_ messages: [[String: Any]]) -> String? {
        for message in messages where message["role"] as? String == "user" {
            let content = messageText(message["content"])
            guard content.hasPrefix("[BENCH:"),
                  let closing = content.firstIndex(of: "]") else { continue }
            return String(content[content.index(content.startIndex, offsetBy: 7)..<closing])
        }
        return nil
    }

    private func messageText(_ content: Any?) -> String {
        if let text = content as? String { return text }
        guard let parts = content as? [[String: Any]] else { return "" }
        return parts.compactMap { part in
            part["type"] as? String == "text" ? part["text"] as? String : nil
        }.joined(separator: "\n")
    }

    private typealias SuccessfulToolCall = (name: String, arguments: [String: Any])

    private func successfulToolCalls(_ messages: [[String: Any]]) -> [SuccessfulToolCall] {
        var successfulIDs = Set<String>()
        for message in messages where message["role"] as? String == "tool" {
            guard let id = message["tool_call_id"] as? String,
                  let data = messageText(message["content"]).data(using: .utf8),
                  let result = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                  result["error"] == nil else { continue }
            successfulIDs.insert(id)
        }
        var result: [SuccessfulToolCall] = []
        for message in messages {
            guard let calls = message["tool_calls"] as? [[String: Any]] else { continue }
            for call in calls {
                guard let id = call["id"] as? String, successfulIDs.contains(id),
                      let function = call["function"] as? [String: Any],
                      let name = function["name"] as? String,
                      let encoded = function["arguments"] as? String,
                      let data = encoded.data(using: .utf8),
                      let arguments = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else { continue }
                result.append((name, arguments))
            }
        }
        return result
    }

    private func hasC1CompletionEvidence(_ calls: [SuccessfulToolCall]) -> Bool {
        let searchedByURL = calls.contains { call in
            guard call.name == "device_invoke",
                  call.arguments["capability"] as? String == "open_url",
                  let params = call.arguments["params"] as? [String: Any],
                  let url = params["url"] as? String,
                  let items = URLComponents(string: url)?.queryItems else { return false }
            return items.contains { ["q", "query", "keyword", "wd"].contains($0.name.lowercased()) }
        }
        let searchedByInput = calls.contains {
            $0.name == "device_act" && $0.arguments["action"] as? String == "input"
        }
        let click = calls.contains {
            $0.name == "device_act" && $0.arguments["action"] as? String == "click"
        }
        let scroll = calls.contains {
            $0.name == "device_act" && ["scroll", "swipe"].contains($0.arguments["action"] as? String)
        }
        return (searchedByURL || searchedByInput) && click && scroll
    }

    private func responseFormat(taskID: String) -> [String: Any] {
        guard taskID == "C1" else { return ["type": "json_object"] }
        let itemSchema: [String: Any] = [
            "type": "object",
            "additionalProperties": false,
            "properties": [
                "title": ["type": "string"],
                "snippet": ["type": "string"],
            ],
            "required": ["title", "snippet"],
        ]
        let detailSchema: [String: Any] = [
            "type": "object",
            "additionalProperties": false,
            "properties": [
                "title": ["type": "string"],
                "summary": ["type": "string"],
            ],
            "required": ["title", "summary"],
        ]
        let schema: [String: Any] = [
            "type": "object",
            "additionalProperties": false,
            "properties": [
                "query": ["type": "string"],
                "items": ["type": "array", "minItems": 5, "maxItems": 5, "items": itemSchema],
                "detail": detailSchema,
                "scrolled": ["type": "boolean"],
            ],
            "required": ["query", "items", "detail", "scrolled"],
        ]
        return [
            "type": "json_schema",
            "json_schema": ["name": "mobile_agent_c1_result", "strict": true, "schema": schema],
        ]
    }
}

File: apps/android/app/src/main/java/ai/mobileagent/agent/KimiClient.kt
package ai.mobileagent.agent

import android.content.Context
import android.util.Log
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.currentCoroutineContext
import kotlinx.coroutines.job
import kotlinx.coroutines.withContext
import org.json.JSONArray
import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URL

class OpenAICompatibleClient(
    private val apiKey: String,
    val endpoint: ModelEndpoint = ModelEndpoint.default,
    private val context: Context? = null,
) {
    @Volatile
    private var activeConnection: HttpURLConnection? = null

    fun cancel() {
        activeConnection?.disconnect()
    }

    suspend fun complete(messages: JSONArray, tools: JSONArray): JSONObject = withContext(Dispatchers.IO) {
        DebugModelDelay.consume(context)
        val startedAt = System.currentTimeMillis()
        val benchmarkTaskId = benchmarkTaskId(messages)
        val successfulToolCalls = successfulToolCalls(messages)
        val responseFormat = benchmarkTaskId
            ?.let { taskId -> BenchmarkResponsePolicy.choose(taskId, successfulToolCalls)?.let { taskId to it } }
            ?.let { (taskId, _) -> responseFormat(taskId) }
        Log.i(
            "MobileAgentPi",
            "model request started; endpointHost=${endpoint.host} model=${endpoint.model} " +
                "messages=${messages.length()} tools=${tools.length()} " +
                "benchmarkTask=${benchmarkTaskId ?: "none"} responseFormat=${responseFormat?.optString("type") ?: "text"}",
        )
        val connection = URL("${endpoint.baseUrl}/chat/completions").openConnection() as HttpURLConnection
        activeConnection = connection
        val requestJob = currentCoroutineContext().job
        val cancellationHandle = requestJob.invokeOnCompletion { cause ->
            if (cause is CancellationException) connection.disconnect()
        }
        try {
            connection.requestMethod = "POST"
            connection.connectTimeout = 30_000
            connection.readTimeout = 180_000
            connection.doOutput = true
            connection.setRequestProperty("Authorization", "Bearer $apiKey")
            connection.setRequestProperty("Content-Type", "application/json")
            val body = JSONObject()
                .put("model", endpoint.model)
                .put("messages", messages)
                .put("tools", tools)
                .put("tool_choice", "auto")
            if (endpoint.sendsReasoningEffort) body.put("reasoning_effort", "low")
            responseFormat?.let { body.put("response_format", it) }
            connection.outputStream.use { it.write(body.toString().toByteArray()) }
            val stream = if (connection.responseCode in 200..299) connection.inputStream else connection.errorStream
            val response = stream.bufferedReader().use { it.readText() }
            Log.i("MobileAgentPi", "model request finished; status=${connection.responseCode} elapsedMs=${System.currentTimeMillis() - startedAt}")
            val json = JSONObject(response)
            if (connection.responseCode !in 200..299) {
                val message = json.optJSONObject("error")?.optString("message") ?: "HTTP ${connection.responseCode}"
                error("MODEL_API_ERROR: $message")
            }
            json
        } catch (error: Throwable) {
            if (requestJob.isCancelled) {
                Log.i("MobileAgentPi", "model request cancelled; elapsedMs=${System.currentTimeMillis() - startedAt}")
                throw CancellationException("MODEL_REQUEST_CANCELLED").also { it.initCause(error) }
            }
            throw error
        } finally {
            cancellationHandle.dispose()
            connection.disconnect()
            if (activeConnection === connection) activeConnection = null
        }
    }

    private fun benchmarkTaskId(messages: JSONArray): String? = (0 until messages.length()).firstNotNullOfOrNull { index ->
        val message = messages.optJSONObject(index) ?: return@firstNotNullOfOrNull null
        if (message.optString("role") != "user") return@firstNotNullOfOrNull null
        BENCHMARK_TASK.find(messageText(message.opt("content")))?.groupValues?.get(1)
    }

    private fun messageText(content: Any?): String = when (content) {
        is String -> content
        is JSONArray -> (0 until content.length()).joinToString("\n") { index ->
            content.optJSONObject(index)?.takeIf { it.optString("type") == "text" }?.optString("text").orEmpty()
        }
        else -> ""
    }

    private fun successfulToolCalls(messages: JSONArray): List<CompletedToolCall> {
        val successfulIds = mutableSetOf<String>()
        for (index in 0 until messages.length()) {
            val message = messages.optJSONObject(index) ?: continue
            if (message.optString("role") != "tool") continue
            val result = runCatching { JSONObject(messageText(message.opt("content"))) }.getOrNull() ?: continue
            if (!result.has("error")) message.optString("tool_call_id").takeIf(String::isNotBlank)?.let(successfulIds::add)
        }
        val calls = mutableListOf<CompletedToolCall>()
        for (index in 0 until messages.length()) {
            val message = messages.optJSONObject(index) ?: continue
            val toolCalls = message.optJSONArray("tool_calls") ?: continue
            for (callIndex in 0 until toolCalls.length()) {
                val call = toolCalls.optJSONObject(callIndex) ?: continue
                if (call.optString("id") !in successfulIds) continue
                val function = call.optJSONObject("function") ?: continue
                val arguments = runCatching { JSONObject(function.optString("arguments")) }.getOrNull() ?: continue
                calls += CompletedToolCall(
                    name = function.optString("name"),
                    action = arguments.optString("action").takeIf(String::isNotBlank),
                    capability = arguments.optString("capability").takeIf(String::isNotBlank),
                    url = arguments.optJSONObject("params")?.optString("url")?.takeIf(String::isNotBlank),
                )
            }
        }
        return calls
    }

    private fun responseFormat(taskId: String): JSONObject {
        if (taskId != "C1") return JSONObject().put("type", "json_object")
        val itemSchema = JSONObject()
            .put("type", "object")
            .put("additionalProperties", false)
            .put("properties", JSONObject()
                .put("title", JSONObject().put("type", "string"))
                .put("snippet", JSONObject().put("type", "string")))
            .put("required", JSONArray().put("title").put("snippet"))
        val detailSchema = JSONObject()
            .put("type", "object")
            .put("additionalProperties", false)
            .put("properties", JSONObject()
                .put("title", JSONObject().put("type", "string"))
                .put("summary", JSONObject().put("type", "string")))
            .put("required", JSONArray().put("title").put("summary"))
        val schema = JSONObject()
            .put("type", "object")
            .put("additionalProperties", false)
            .put("properties", JSONObject()
                .put("query", JSONObject().put("type", "string"))
                .put("items", JSONObject()
                    .put("type", "array")
                    .put("minItems", 5)
                    .put("maxItems", 5)
                    .put("items", itemSchema))
                .put("detail", detailSchema)
                .put("scrolled", JSONObject().put("type", "boolean")))
            .put("required", JSONArray().put("query").put("items").put("detail").put("scrolled"))
        return JSONObject()
            .put("type", "json_schema")
            .put("json_schema", JSONObject()
                .put("name", "mobile_agent_c1_result")
                .put("strict", true)
                .put("schema", schema))
    }

    companion object {
        private val BENCHMARK_TASK = Regex("^\\[BENCH:([A-Z][0-9]+)]")
    }
}
